## **Исследование данных о продаже подержанных автомобилей на eBay Kleinanzeigen**

В данном проекте исследуются данные об объявлениях по продаже автомобилей с популярной площадки eBay Kleinanzeigen. Основная цель анализа заключается в выявлении факторов, влияющих на стоимость автомобиля, таких как марка, модель, пробег и год регистрации.

В ходе работы будут изучены распределения основных признаков, выполнена очистка данных и проведён исследовательский анализ. Это позволит выявить характерные особенности рынка подержанных автомобилей и определить взаимосвязи между характеристиками автомобилей и их стоимостью.

In [1]:
# Импортируем необходимые библиотеки
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'latin1')

In [3]:
autos

,dateCrawled,name,seller,offerType,price,abtest,vehicleType,yearOfRegistration,gearbox,powerPS,model,odometer,monthOfRegistration,fuelType,brand,notRepairedDamage,dateCreated,nrOfPictures,postalCode,lastSeen
0,2016-03-26 17:47:46,Peugeot_807_160_NAVTECH_ON_BOARD,privat,Angebot,"$5,000",control,bus,2004,manuell,158,andere,"150,000km",3,lpg,peugeot,nein,2016-03-26 00:00:00,0,79588,2016-04-06 06:45:54
1,2016-04-04 13:38:56,BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik,privat,Angebot,"$8,500",control,limousine,1997,automatik,286,7er,"150,000km",6,benzin,bmw,nein,2016-04-04 00:00:00,0,71034,2016-04-06 14:45:08
2,2016-03-26 18:57:24,Volkswagen_Golf_1.6_United,privat,Angebot,"$8,990",test,limousine,2009,manuell,102,golf,"70,000km",7,benzin,volkswagen,nein,2016-03-26 00:00:00,0,35394,2016-04-06 20:15:37
3,2016-03-12 16:58:10,Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...,privat,Angebot,"$4,350",control,kleinwagen,2007,automatik,71,fortwo,"70,000km",6,benzin,smart,nein,2016-03-12 00:00:00,0,33729,2016-03-15 03:16:28
4,2016-04-01 14:38:50,Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...,privat,Angebot,"$1,350",test,kombi,2003,manuell,0,focus,"150,000km",7,benzin,ford,nein,2016-04-01 00:00:00,0,39218,2016-04-01 14:38:50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
49995,2016-03-27 14:38:19,Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon,privat,Angebot,"$24,900",control,limousine,2011,automatik,239,q5,"100,000km",1,diesel,audi,nein,2016-03-27 00:00:00,0,82131,2016-04-01 13:47:40
49996,2016-03-28 10:50:25,Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...,privat,Angebot,"$1,980",control,cabrio,1996,manuell,75,astra,"150,000km",5,benzin,opel,nein,2016-03-28 00:00:00,0,44807,2016-04-02 14:18:02
49997,2016-04-02 14:44:48,Fiat_500_C_1.2_Dualogic_Lounge,privat,Angebot,"$13,200",test,cabrio,2014,automatik,69,500,"5,000km",11,benzin,fiat,nein,2016-04-02 00:00:00,0,73430,2016-04-04 11:47:27
49998,2016-03-08 19:25:42,Audi_A3_2.0_TDI_Sportback_Ambition,privat,Angebot,"$22,900",control,kombi,2013,manuell,150,a3,"40,000km",11,diesel,audi,nein,2016-03-08 00:00:00,0,35683,2016-04-05 16:45:07


Изучим данный датасет более подробно:

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Исходя из первичного анализа можно заметить, что значения в столбцах `price` и `odometer` имеют тип данных object. В дальнейшем их необходимо преобразовать к типу int.

Также столбцы `vehicleType`, `gearbox`, `model`, `fuelType` и `notRepairedDamage` содержат пропущенные значения (null).

---

#### Для начала внесём изменения в названия столбцов, чтобы повысить их читаемость и упростить дальнейшую работу с данными:

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 
                 'price', 'ab_test', 'vehicle_type', 'registration_year', 
                 'gearbox', 'power_ps', 'model', 'odometer', 
                 'registration_month', 'fuel_type', 'brand','unrepaired_damage', 
                 'ad_created', 'num_photos', 'postal_code','last_seen']

In [7]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

---

#### Как было отмечено ранее, в датасете присутствуют признаки с некорректно определёнными типами данных (`price`, `odometer`). Выполним преобразование этих признаков:

In [8]:
print(autos['price'].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


Значения в столбце `price` представлены в виде строк с разделителем тысяч (`,`), поэтому перед преобразованием типа данных к `int` запятые необходимо удалить:

In [9]:
autos['price'] = autos['price'].str.replace('$', '', regex=False)
autos['price'] = autos['price'].str.replace(',', '')
autos['price'] = autos['price'].astype(int)

Аналогично для `odometer`:

In [10]:
print(autos['odometer'].unique())

['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [11]:
autos['odometer'] = autos['odometer'].str.split('k').str[0]
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].astype(int)

Для повышения информативности данных внесём изменения в названия столбцов:

In [12]:
autos.rename({
    'price':'price_usd',
    'odometer':'odometer_km'
}, axis = 1, inplace = True)

---

#### Проанализируем основные статистические характеристики данных с помощью метода `describe()`:

In [13]:
autos.describe() # Для числовых признаков

,price_usd,registration_year,power_ps,odometer_km,registration_month,num_photos,postal_code
count,5.000000e+04,50000.000000,50000.000000,50000.000000,50000.000000,50000.0,50000.000000
mean,9.840044e+03,2005.073280,116.355920,125732.700000,5.723360,0.0,50813.627300
std,4.811044e+05,105.712813,209.216627,40042.211706,3.711984,0.0,25779.747957
min,0.000000e+00,1000.000000,0.000000,5000.000000,0.000000,0.0,1067.000000
25%,1.100000e+03,1999.000000,70.000000,125000.000000,3.000000,0.0,30451.000000
50%,2.950000e+03,2003.000000,105.000000,150000.000000,6.000000,0.0,49577.000000
75%,7.200000e+03,2008.000000,150.000000,150000.000000,9.000000,0.0,71540.000000
max,1.000000e+08,9999.000000,17700.000000,150000.000000,12.000000,0.0,99998.000000


Все значения в столбце `num_photos` равны 0.

In [14]:
autos.describe(include = ['O']) # Для нечисловых признаков

,date_crawled,name,seller,offer_type,ab_test,vehicle_type,gearbox,model,fuel_type,brand,unrepaired_damage,ad_created,last_seen
count,50000,50000,50000,50000,50000,44905,47320,47242,45518,50000,40171,50000,50000
unique,48213,38754,2,2,2,8,2,245,7,40,2,76,39481
top,2016-04-02 11:37:04,Ford_Fiesta,privat,Angebot,test,limousine,manuell,golf,benzin,volkswagen,nein,2016-04-03 00:00:00,2016-04-07 06:17:27
freq,3,78,49999,49999,25756,12859,36993,4024,30107,10687,35232,1946,8


Столбцы `seller` и `offer_type` практически константны: в каждом из них одно значение встречается в 49 999 случаях из 50 000.

---

#### Проведём более детальный анализ признаков `price_usd` и `odometer_km` на наличие выбросов:

In [15]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [16]:
# Самые высокие значения
autos['price_usd'].value_counts().sort_index(ascending = False).head(12)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
Name: price_usd, dtype: int64

In [17]:
# Самые низкие значения
autos['price_usd'].value_counts().sort_index(ascending = True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price_usd, dtype: int64

Можно заметить, что в столбце `price_usd` присутствуют аномальные значения. С одной стороны, встречаются автомобили с ценой, равной 0 долларов, что, вероятно, связано с ошибками ввода данных или отсутствием указанной стоимости. С другой стороны, наблюдаются значения, превышающие 1 млн долларов, которые являются крайне нетипичными для рассматриваемой выборки и могут существенно искажать результаты анализа. В связи с этим такие наблюдения целесообразно исключить из набора данных для повышения его репрезентативности и качества последующего анализа.

Дополнительно исключим столбцы `num_photos`, `seller` и `offer_type`, так как они содержат только одно уникальное значение и не несут полезной информации для дальнейшего анализа.

In [18]:
autos = autos.loc[autos['price_usd'].between(1, 350000), ['date_crawled', 'name', 'price_usd', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code','last_seen']]

Аналогично для столбца `odometer_km`:

In [19]:
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos['odometer_km'].value_counts().sort_index(ascending = True).head(20)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

Анализ столбца `odometer_km` не выявил критических выбросов или некорректных значений. В связи с этим дополнительная обработка данного признака не требуется.

---

#### Теперь перейдём к анализу столбцов, содержащих даты (`ad_created`, `date_crawled`, `registration_year`), и определим временной диапазон, который охватывают данные:

- Для  `ad_created` (дата создания объявления владельцем):

In [21]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

Временной диапазон в столбце `ad_created` охватывает период с `2015-06-11` по `2016-04-07`.

In [22]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending=False).head(20)

2016-04-03    0.038855
2016-03-20    0.037949
2016-03-21    0.037579
2016-04-04    0.036858
2016-03-12    0.036755
2016-03-14    0.035190
2016-04-02    0.035149
2016-03-28    0.034984
2016-03-07    0.034737
2016-03-29    0.034037
2016-03-15    0.034016
2016-04-01    0.033687
2016-03-19    0.033687
2016-03-30    0.033501
2016-03-08    0.033316
2016-03-09    0.033151
2016-03-11    0.032904
2016-03-22    0.032801
2016-03-26    0.032266
2016-03-23    0.032060
Name: ad_created, dtype: float64

Также по распределению дат видно, что большинство объявлений были опубликованы именно в период с марта по апрель 2016 года.

- Для  `date_crawled` (дата, когда объявление было впервые зафиксировано веб-краулером сайта):

In [23]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

Временной диапазон в столбце `date_crawled` охватывает период с `2016-03-05` по `2016-04-07`. 

- Для  `registration_year` (год первой регистрации автомобиля):

In [24]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [25]:
autos['registration_year'].value_counts().sort_index(ascending=True).head(10)

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
Name: registration_year, dtype: int64

Минимальные значения года регистрации автомобиля (в диапазоне `1000–1800`) являются явной аномалией. Максимальный год регистрации не может превышать `2016`, поскольку временной диапазон в столбце `ad_created` (дата создания объявления) охватывает период с `2015-06-11` по `2016-04-07`. Следовательно, наблюдения с некорректными значениями года регистрации подлежат дополнительной проверке и последующему исключению из анализа.

Посчитаем количество записей с некорректными значениями года регистрации автомобиля:

In [26]:
autos[~autos['registration_year'].between(1910, 2016)].shape[0]

1884

Выявлено 1884 записи с некорректными значениями года регистрации автомобиля (меньше 1910 и больше 2016). Данные наблюдения подлежат исключению из дальнейшего анализа:

In [27]:
autos = autos.loc[autos['registration_year'].between(1910, 2016), :]

Рассмотрим распределение оставшихся значений:

In [28]:
autos['registration_year'].value_counts(normalize=True).head(25)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007926
Name: registration_year, dtype: float64

#### **Вывод**:

Большинство автомобилей зарегистрированы примерно в начале-середине 2000-х годов, с разбросом в сторону более ранних значений, включая начало 1990-х.

Учитывая, что объявления публиковались в период 2015-2016 годов, можно отметить, что основная масса автомобилей в выборке имеет возраст порядка 10-20 лет. Полученные значения выглядят реалистично и соответствуют ожидаемой структуре рынка подержанных автомобилей.

Винтажные и очень старые автомобили встречаются редко и не оказывают существенного влияния на общее распределение данных.

---

#### Рассмотрим распределение средних цен автомобилей в зависимости от их марки (`brand`):

Ограничимся топ-10 наиболее распространёнными брендами:

In [29]:
brand_price_dict = {}

for a in autos['brand'].value_counts().index[:10]:
    selected_rows = autos.loc[autos['brand'] == a, 'price_usd']
    avg_price = selected_rows.mean()
    brand_price_dict[a] = avg_price

for key, value in brand_price_dict.items():
    print(f"{key}: ${round(value)}")

volkswagen: $5402
bmw: $8333
opel: $2975
mercedes_benz: $8628
audi: $9337
ford: $3749
renault: $2475
peugeot: $3094
fiat: $2814
seat: $4397


#### **Вывод**:

Среди 10 наиболее распространённых марок автомобилей наибольшая средняя цена наблюдается у премиальных брендов. Лидерами по стоимости являются `Audi ($9337)`, `Mercedes-Benz ($8628)` и `BMW ($8333)`. Это ожидаемый результат, поскольку данные марки относятся к премиальному сегменту рынка.

---

#### Для каждого из этих брендов вычислим среднее значение пробега (`odometer_km`):

In [30]:
brand_mileage_dict = {}

for a in autos['brand'].value_counts().index[:10]:
    selected_rows = autos.loc[autos['brand'] == a, 'odometer_km']
    avg_mileage = selected_rows.mean()
    brand_mileage_dict[a] = avg_mileage

for key, value in brand_mileage_dict.items():
    print(f"{key}: {round(value)} mileage")

volkswagen: 128707 mileage
bmw: 132573 mileage
opel: 129310 mileage
mercedes_benz: 130788 mileage
audi: 129157 mileage
ford: 124266 mileage
renault: 128071 mileage
peugeot: 127154 mileage
fiat: 117122 mileage
seat: 121131 mileage


Для дальнейшего сравнения преобразуем словари в объекты pandas Series и объединим их в единый DataFrame для анализа средних значений пробега и цены по каждому бренду:

In [31]:
brand_price_series = pd.Series(brand_price_dict)
brand_mileage_series = pd.Series(brand_mileage_dict)

df_brand_info = pd.DataFrame(brand_price_series, columns=['price_usd'])
df_brand_info['odometer_km'] = brand_mileage_series

In [32]:
df_brand_info.sort_values('price_usd', ascending = False)

,price_usd,odometer_km
audi,9336.687454,129157.386785
mercedes_benz,8628.450366,130788.363313
bmw,8332.820518,132572.513140
volkswagen,5402.410262,128707.158791
seat,4397.230950,121131.301290
ford,3749.469507,124266.012872
peugeot,3094.017229,127153.625269
opel,2975.241935,129310.035842
fiat,2813.748538,117121.971596
renault,2474.864607,128071.331213


#### **Вывод**:

Несмотря на то, что автомобили марок `Audi ($9337)`, `Mercedes-Benz ($8628)` и `BMW ($8333)` имеют самые высокие средние цены, их средний пробег также находится среди наибольших в рассматриваемой группе и составляет около 129-133 тыс. км.

Это свидетельствует о том, что высокая стоимость автомобилей данных брендов не объясняется исключительно меньшим пробегом. Вероятно, на цену оказывают влияние и другие факторы, такие как класс автомобиля, технические характеристики, уровень оснащения, репутация бренда и восприятие его ценности на вторичном рынке.

Таким образом, в рассматриваемой выборке не наблюдается прямой зависимости между пробегом и стоимостью автомобиля: меньший пробег не обязательно соответствует более высокой цене.

---

#### Рассмотрим наиболее распространённые комбинации марки (`brand`) и модели (`model`) автомобилей:

In [33]:
autos[['brand','model']].value_counts().head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
dtype: int64

#### **Вывод**:

Анализ наиболее распространённых сочетаний марки и модели автомобиля показал, что наибольшее количество объявлений приходится на несколько популярных моделей различных ценовых сегментов. Абсолютным лидером является `Volkswagen Golf` (3707 объявлений), за которым следуют `BMW 3er` (2615 объявлений), `Volkswagen Polo` (1609 объявлений) и `Opel Corsa` (1592 объявления).

Среди наиболее популярных моделей отдельных брендов можно выделить:

- Volkswagen: Golf, Polo, Passat;
- BMW: 3er, 5er;
- Opel: Corsa, Astra;
- Mercedes-Benz: C-Klasse, E-Klasse;
- Audi: A4.